pkg/controller/infrastructureconfig/validations.go

```diff
@@ -0,0 +1,98 @@
+package infrastructureconfig
+
+import (
+	"fmt"
+	configv1 "github.com/openshift/api/config/v1"
+	"github.com/openshift/cluster-network-operator/pkg/util/ip"
+	utilnet "k8s.io/utils/net"
+	"net/netip"
+)
+
+// validateVipsWithVips compares API and Ingress VIPs to detect whether those can coexist. It is supposed to catch
+// scenarios when configuring particular set of addresses is valid in the given environment (by machine networks)
+// but invalid between set of VIPs. In order to be valid, VIPs need to fulfil the following spec:
+//
+//   - 0 VIPs is allowed as a migration from pre-4.16 (where Spec is empty)
+//   - the number of VIPs is either 1 or 2
+//   - number of VIPs needs to be the same for API and ingress
+//   - no any pair of VIPs can be equal, unless External Load Balancer is used
+//   - IP stack of the second VIP needs to be different from IP stack of the first VIP
+func validateVipsWithVips(api, ingress []configv1.IP, elb bool) error {
+	if len(api) != 0 || len(ingress) != 0 {
+		if len(api) != 1 && len(api) != 2 {
+			return fmt.Errorf("number of API VIPs needs to be 1 or 2, got %d", len(api))
+		}
+		if len(ingress) != 1 && len(ingress) != 2 {
+			return fmt.Errorf("number of Ingress VIPs needs to be 1 or 2, got %d", len(ingress))
+		}
+	}
+	if len(api) != len(ingress) {
+		return fmt.Errorf("number of API VIPs (%d) does not match number of Ingress VIPs (%d)", len(api), len(ingress))
+	}
+
+	// For external load balancer we allow VIPs to be equal.
+	if !elb {
+		for i := 0; i < len(api); i++ {
+			if api[i] == ingress[i] {
+				return fmt.Errorf("VIPs cannot be equal, got '%s' for API and '%s' for Ingress", api[i], ingress[i])
+			}
+		}
+	}
+
+	if len(api) > 1 {
+		ok, err := utilnet.IsDualStackIPStrings(ip.IPsToStrings(api))
+		if !ok || err != nil {
+			return fmt.Errorf("with more than 1 API VIP at least one from each IP family is required")
+		}
+
+		ok, err = utilnet.IsDualStackIPStrings(ip.IPsToStrings(ingress))
+		if !ok || err != nil {
+			return fmt.Errorf("with more than 1 Ingress VIP at least one from each IP family is required")
+		}
+	}
+
+	return nil
+}
+
+// validateVipsWithMachineNetworks compares VIPs and machine networks to detect whether those can coexist. It is
+// supposed to catch scenarios when configuring particular set of VIPs would be illegal given the provided Machine
+// Networks. In order to be valid, VIPs need to fulfil the following spec:
+//
+//   - IP stack of the first VIP needs to be IP stack of the first machine network
```

Comment:
Yes, and probably never will be (because there is no need). The logic behind is as follows

* we do not allow changing the first VIP nor the first Machine Network
* when you create a cluster, API and installer make sure that 1st VIP matches 1st Machine Network

so from points above we can guarantee that if you ever got a cluster to install correctly (up to the point when CNO runs), the stacks match. It would start making sense to allow those modifications if K8s ever allow conversions from IPv4-only to IPv6-only but AFAIK this is not happening

Reply:
For the sake of completeness I added a comment explaining that this rule applies but we don't check it in CNO